File: HouseRobber.java
/*
Runtime: 0 ms, faster than 100.00% of Java online submissions for House Robber.
Memory Usage: 39.5 MB, less than 95.96% of Java online submissions for House Robber.
*/
// Optimized Solution : TC: O(n) SC: O(n)
class Solution {
    public int rob(int[] nums) {
        
        // Standard Dynamic Programming Procedure With An Array Of Size nums.length + 1
        int[] dp = new int[nums.length + 1];
        
        // Assigning The first 2 elements Of The DP Array
        dp[0] = 0;
        dp[1] = nums[0];
        
        // Calculating The dp Element Value As Per:
        // Math.max Of Previous Element Of dp Array And Sum Of Current nums Element + The Last Non Adjacent Element Sum
        for(int i=2; i<dp.length; i++){
            dp[i] = Math.max(dp[i-1], nums[i-1] + dp[i-2]);
        }
        
        // Returning The Final Result Which Is Stored In The Last Element Of The dp Array
        return dp[dp.length - 1];
        
    }
}

File: HouseRobber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// TC: O(N) SC: O(1) 
var rob = function(nums) {
    if (nums.length <= 2) return Math.max(...nums); // Handle base cases

    let prev1 = nums[0]; // Initialize prev1: Maximum amount robbable upto the previous house
    let prev2 = Math.max(nums[0], nums[1]); // Initialize prev2: Maximum amount robbable upto the second previous house 

    for (let i = 2; i < nums.length; i++) { 
        const temp = prev2;  // Store the previous 'prev2' to not lose track of its value
        prev2 = Math.max(prev2, nums[i] + prev1); // Update 'prev2': Include the current house or not?
        prev1 = temp; // Update 'prev1': 'temp' holds the old value of the new 'prev2'  
    }

    return prev2; // Return the final result (maximum amount robbable)
};

// TC: O(N) SC: O(1) Dynamic Programming
var rob = function(nums) {
    if(nums.length <=2) return Math.max(...nums);
    const dp = [nums[0], Math.max(nums[0], nums[1])];
    for(let i=2; i<nums.length; i++) {
        dp[i] = Math.max(dp[i - 1], nums[i] + dp[i - 2]);
    }
    return dp.pop();
};
